lib/chef/resource/windows_certificate.rb

```diff
@@ -161,35 +161,69 @@ class WindowsCertificate < Chef::Resource
       end
 
       action_class do
+
+        CERT_SYSTEM_STORE_LOCAL_MACHINE                    = 0x00020000
+        CERT_SYSTEM_STORE_CURRENT_USER                     = 0x00010000
+
         def add_cert(cert_obj)
-          store = ::Win32::Certstore.open(new_resource.store_name)
+          store = ::Win32::Certstore.open(new_resource.store_name, store_location: get_cert_location[1])
           store.add(cert_obj)
         end
 
         def add_pfx_cert
           exportable = new_resource.exportable ? 1 : 0
-          store = ::Win32::Certstore.open(new_resource.store_name)
+          store = ::Win32::Certstore.open(new_resource.store_name, store_location: get_cert_location[1])
           store.add_pfx(new_resource.source, new_resource.pfx_password, exportable)
         end
 
         def delete_cert
           store = ::Win32::Certstore.open(new_resource.store_name)
-          store.delete(new_resource.source)
+          if check_string(new_resource.source) && new_resource.source.length == 40
+            store.delete(new_resource.source)
+          else
+            thumbprint = powershell_exec!(get_thumbprint(new_resource.store_name, get_cert_location[0], new_resource.source)).result
+            store.delete(thumbprint)
+          end
```

Comment:
You should extract a helper method out of this so that all three of these similar blocks of code get extracted...

```
store.delete(thumbprintcode(new_resource.source))
store.get(thumbprintcode(new_resource.source))
store.valid?(thumbprintcode(thumbprint))
```
with:

```
def thumbprintcode(thumbprint)
  if check_string(thumbprint) && thumbprint.length == 40
    store.get(thumbprint)
  else
    thumbprint = powershell_exec!(get_thumbprint(new_resource.store_name, get_cert_location[0], new_resource.source)).result
    store.get(thumbprint)
  end
end
```

or:

```
def thumbprintcode(thumbprint)
  if !check_string(thumbprint) || thumbprint.length != 40 
    thumbprint = powershell_exec!(get_thumbprint(new_resource.store_name, get_cert_location[0], new_resource.source)).result
  end
  store.get(thumbprint)
end
```

you could also invert check_string, add the length != 40 check to it and turn it into `def invalid_thumbprintcode?` and then you'd have:

```
def thumbprintcode(thumbprint)
  if invalid_thumbprintcode?(thumbprint)
    thumbprint = powershell_exec!(get_thumbprint(new_resource.store_name, get_cert_location[0], new_resource.source)).result
  end
  store.get(thumbprint)
end
```

With that method now looking like:

```
def invalid_thumbprintcode?(code)
  !(code.scan(/\H/).empty? && code.length == 40)
end
```